cmd/transcriber/transcribe/text.go

```diff
@@ -0,0 +1,143 @@
+}
+
+func (o *TextCompactOptions) SetDefaults() {
+	o.SilenceThresholdMs = 2000
+	o.MaxSegmentDurationMs = 10000
+}
+
+func (o *TextCompactOptions) IsEmpty() bool {
+	return o == nil || *o == TextCompactOptions{}
+}
+
+type TextOptions struct {
+	CompactOptions TextCompactOptions
+}
+
+func (o *TextOptions) SetDefaults() {
+	o.CompactOptions.SetDefaults()
+}
+
+func (o *TextOptions) IsValid() error {
+	if o.CompactOptions.SilenceThresholdMs <= 0 {
+		return fmt.Errorf("SilenceThresholdMs should be a positive number")
+	}
+
+	if o.CompactOptions.MaxSegmentDurationMs <= 0 {
+		return fmt.Errorf("MaxSegmentDurationMs should be a positive number")
+	}
+
+	return nil
+}
+
+func (o *TextOptions) IsEmpty() bool {
+	return o.CompactOptions.IsEmpty()
+}
+
+func (o *TextOptions) ToEnv() []string {
+	return []string{
+		fmt.Sprintf("TEXT_COMPACT_SILENCE_THRESHOLD_MS=%d", o.CompactOptions.SilenceThresholdMs),
+		fmt.Sprintf("TEXT_COMPACT_MAX_SEGMENT_DURATION_MS=%d", o.CompactOptions.MaxSegmentDurationMs),
+	}
+}
+
+func (o *TextOptions) FromEnv() {
+	o.CompactOptions.SilenceThresholdMs, _ = strconv.Atoi(os.Getenv("TEXT_COMPACT_SILENCE_THRESHOLD_MS"))
+	o.CompactOptions.MaxSegmentDurationMs, _ = strconv.Atoi(os.Getenv("TEXT_COMPACT_MAX_SEGMENT_DURATION_MS"))
+}
+
+func (o *TextOptions) ToMap() map[string]any {
+	return map[string]any{
+		"text_compact_silence_threshold_ms":    o.CompactOptions.SilenceThresholdMs,
+		"text_compact_max_segment_duration_ms": o.CompactOptions.MaxSegmentDurationMs,
+	}
+}
+
+func (o *TextOptions) FromMap(m map[string]any) {
+	// These can either be int or float64 dependning whether they have been
+	// previously marshaled or not.
+	switch m["text_compact_silence_threshold_ms"].(type) {
+	case int:
+		o.CompactOptions.SilenceThresholdMs = m["text_compact_silence_threshold_ms"].(int)
+	case float64:
+		o.CompactOptions.SilenceThresholdMs = int(m["text_compact_silence_threshold_ms"].(float64))
+	}
+
+	switch m["text_compact_max_segment_duration_ms"].(type) {
+	case int:
+		o.CompactOptions.MaxSegmentDurationMs = m["text_compact_max_segment_duration_ms"].(int)
+	case float64:
+		o.CompactOptions.MaxSegmentDurationMs = int(m["text_compact_max_segment_duration_ms"].(float64))
+	}
+}
+
+func compactSegments(segments []namedSegment, opts TextCompactOptions) []namedSegment {
+	if len(segments) < 2 {
+		return segments
+	}
+
+	out := []namedSegment{segments[0]}
+
+	for i := 1; i < len(segments); i++ {
+		currSeg := segments[i]
+		prevSeg := segments[i-1]
+
+		// We join the segments if:
+		// - The speaker hasn't changed. This is required to guarantee order (e.g. question/answer sequences).
+		// - There's less than silenceThresholdMs of pause between the end of a previous text segment and the start of the next one.
+		// - The overall (running) duration of the joined segments is less than maxDurationMs seconds.
+		if currSeg.Speaker == prevSeg.Speaker &&
+			int(currSeg.StartTS-prevSeg.EndTS) < opts.SilenceThresholdMs &&
```

Comment:
Could these be `out[len(out)-1]` as well (if only to avoid switching mental contexts during the if conditions)?

Reply:
Sure, when I started to implement this I didn't realize I would have had to access by index to modify but since we do then may as well keep it consistent.

Reply:
To be noted, those would still refer to `segments`, not `out`.

Reply:
@cpoile If you can confirm on the above I'll apply the changes, just want to make sure you understand the difference.

`prevSeg` would be `segments[i-1]` and `currSeg` would be `segments[i]`.

Reply:
Hm, on second thought, I don't think the change is as helpful, since it's in the context of segments.